perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -176,26 +187,24 @@ protected void deleteAnyGroup(PerunSession sess, Group group, boolean forceDelet
 				// delete all Groups reserved logins from DB
 				getGroupsManagerImpl().deleteGroupReservedLogins(sess, group);
 
-				// Group applications, submitted data and app_form are deleted on cascade with "deleteGroup()"
+				// 1. remove all relations with group g as an operand group.
+				// this removes all relations that depend on this group
+				List<Integer> relations = groupsManagerImpl.getGroupRelations(sess, g.getId());
+				for (Integer groupId : relations) {
+					removeGroupUnion(sess, groupsManagerImpl.getGroupById(sess, groupId), g, true);
+				}
 
+				// 2. remove all relations with group as a result group
+				// We can remove relations without recalculation (@see processRelationMembers) 
+				// because all dependencies of group were deleted in step 1.
+				groupsManagerImpl.removeResultGroupRelations(sess, g);
+
+				// Group applications, submitted data and app_form are deleted on cascade with "deleteGroup()"
 				List<Member> membersFromDeletedGroup = getGroupMembers(sess, g);
+				// Deletes also all direct and indirect members of the group
 				getGroupsManagerImpl().deleteGroup(sess, vo, g);
-
-				Integer parentGroupId = g.getParentGroupId();
-				while(parentGroupId != null) {
-					Group parentGroup;
-					try {
-						parentGroup = getGroupById(sess, parentGroupId);
-					} catch (GroupNotExistsException ex) {
-						throw new ConsistencyErrorException(ex);
-					}
-					List<Member> membersFromParentGroup = getGroupMembers(sess, parentGroup);
```

Comment:
Please add comment. There are actual members of parentGroup after removing members from subGroup.